lib/mod_movements/handler.js

```diff
@@ -0,0 +1,170 @@
+  return new Vec3(delta[0], delta[1], delta[2])
+}
+
+function radialSymmetry (delta) {
+  return [
+    [delta[0], delta[1], delta[2]],
+    [-delta[2], delta[1], delta[0]],
+    [-delta[0], delta[1], -delta[2]],
+    [delta[2], delta[1], -delta[0]]
+  ]
+}
+
+/**
+ * A parser for reading movement plugin files to generate available
+ * movement lists while pathfinding.
+ */
+class ModularMovementsHandler {
+  /**
+   * Creates a new modular movement handler instance.
+   *
+   * @param {boolean} loadDefault - Whether or not to load the default
+   * movement plugin.
+   */
+  constructor (loadDefault = true) {
+    this.movementPlugins = []
+    this.settings = require('../../movements/settings.json')
+
+    if (loadDefault) {
+      this.loadMovements(require('../../movements/defaultMovements.json'))
+    }
+  }
+
+  /**
+   * Loads a new movement plugin to this handler from a Json file.
+   *
+   * @param {*} movements - The movement Json file to load movement
+   * data from.
+   */
+  loadMovements (movements) {
+    this.movementPlugins.push(movements)
+  }
+
+  /**
+   * Gets all available movements (or neighbor nodes) which can be preformed
+   * from the given node.
+   *
+   * @param {*} node - The node to handle.
+   * @param {*} world - The virtual world state.
+   */
+  getNeighbors (node, world) {
+    const neighbors = []
+    for (const plugin of this.movementPlugins) {
+      for (const movement of plugin.movements) {
+        if (movement.symmetry === 'radial') {
+          for (let i = 0; i < 4; i++) {
+            const m = this._handleMovement(movement, i, node, world)
+            if (m) neighbors.push(m)
+          }
+        } else {
+          const m = this._handleMovement(movement, 0, node, world)
+          if (m) neighbors.push(m)
+        }
+      }
+    }
+
+    return neighbors
+  }
+
+  /**
+   * Returns a new movement node if this current movement is considered
+   * valid.
+   *
+   * @param {*} movement - The raw movement to test for validity.
+   * @param {*} node - The node currently being handled.
+   * @param {*} world - The virtual world state.
+   */
+  _handleMovement (movement, direction, node, world) {
+    const hash = this._generateHash(movement.hash, node.position)
+    const m = new Movement(node.position, hash)
+
+    for (const action of movement.actions) {
+      const type = action.type
+      const delta = deltaToVec(radialSymmetry(action.delta)[direction])
+      const requirements = action.requires || []
+
+      if (!this._meetsRequirements(node, world, requirements)) continue
+
+      const block = world.virtualBlockAt(node, node.position + delta)
+      const a = new Action(type, block.position)
+
+      if (type === 'mine') {
+        if (!this._canMine(block)) continue
+
+        a.tool = 'hand' // TODO Determine best tool
+      }
+
+      if (type === 'build') {
+        if (!this._canPlace(block)) continue
+
+        a.scaffolding = 'dirt' // TODO Determine correct scaffolding
```

Comment:
Same comment as for the digging, you don't need to store that information in the action. You only want to know if you have scaffolding blocks remaining, you don't care which one.